setup.py

```diff
@@ -1,25 +1,96 @@
-  sys.exit()
-
-module1 = Extension('ale_python_interface.ale_c_wrapper',
-                    libraries = ['ale_c'],
-                    include_dirs = ['src'],
-                    library_dirs = ['ale_python_interface'],
-                    extra_compile_args=['-D__STDC_CONSTANT_MACROS', '-std=c++11'],
-                    sources=['ale_python_interface/ale_c_wrapper.cpp'])
-setup(name = 'ale_python_interface',
-      version='0.6',
-      description = 'Arcade Learning Environment Python Interface',
-      url='https://github.com/bbitmaster/ale_python_interface',
-      author='Ben Goodrich',
-      license='GPL',
-      ext_modules = [module1],
-      packages=['ale_python_interface'],
-      package_dir={'ale_python_interface': 'ale_python_interface'},
-      package_data={'ale_python_interface': ['libale_c.so']})
+from setuptools import setup, Extension
+from setuptools.command.build_ext import build_ext
+import subprocess
+import platform
+import os
+
+
+class CMakeExtension(Extension):
+    def __init__(self, name, sourcedir="", config=[]):
+        Extension.__init__(self, name, sources=[])
+        self.sourcedir = os.path.abspath(sourcedir)
+        self.config = config
+
+
+class CMakeBuild(build_ext):
+    def build_extensions(self):
+        try:
+            subprocess.check_output(["cmake", "--version"])
+        except OSError:
+            raise RuntimeError(
+                "CMake must be installed to build the extensions: %s"
+                % ", ".join(ext.name for ext in self.extensions)
+            )
+
+        for ext in self.extensions:
+            extdir = os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.name)))
+            cfg = "Debug" if self.debug else "Release"
+
+            cmake_config_args = [
+                "-DOUTPUT_NAME={}".format(ext.name),
+                "-DCMAKE_BUILD_TYPE={}".format(cfg),
+                "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_{}={}".format(cfg.upper(), extdir),
+                "-DCMAKE_RUNTIME_OUTPUT_DIRECTORY_{}={}".format(cfg.upper(), extdir),
+                "-DCMAKE_ARCHIVE_OUTPUT_DIRECTORY_{}={}".format(
+                    cfg.upper(), self.build_temp
+                ),
+            ] + ext.config
+
+            cmake_build_args = []
+
+            # -DCMAKE_BUILD_TYPE doesn't work on Windows
+            # we need to specify --config Release at build type
+            if platform.system() == "Windows":
+                cmake_build_args += ["--config", cfg]
+
+            if not os.path.exists(self.build_temp):
+                os.makedirs(self.build_temp)
+
+            subprocess.check_call(
+                ["cmake", ext.sourcedir] + cmake_config_args, cwd=self.build_temp
+            )
+            subprocess.check_call(
+                ["cmake", "--build", "."] + cmake_build_args, cwd=self.build_temp
+            )
+
+
+with open("README.md", "r") as f:
+    long_description = f.read()
+
+setup(
+    name="ale-py",
+    version="0.6.0",
```

Comment:
Should this be a new version number?

Reply:
Yes, I still have to figure out a better way to manage versions. Right now I do deployments on tagging. You had mentioned this might cause some issues.

Reply:
I think you suggested a good check to avoid forgetting to update the number.  The only remaining downside is that if you tag a release before tests pass, you may have to re-tag it.  You can just wait until tests pass before tagging though.

Reply:
Should this be a new version number?

Reply:
I have some additional work on versioning that I plan on pushing to this PR. The plan is to extract the version from `CMakeLists.txt` so we can have one source of truth for the version info.

Reply:
Versioning `ale-py` works as follows:

The version should be bumped in `CMakeLists.txt`. `setup.py` will parse out the version and append a suffix of `dev` if we aren't running in CI. If we are running in CI and the tagged version matches the parsed version from `CMakeLists.txt` the suffix will be removed and setup will continue.

There is also a check that both the parsed and tagged versions conform to semver rules. This is mainly due to PyPi being picky so we might as well catch it before the rest of the deployment automation continues.